{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bioinformatics tools and training materials knowledge graph \n",
    "### BioSchemas to bridge Bio.Tools and TeSS \n",
    "\n",
    "**Issue and objective**: TeSS training materials are not always annotated with the EDAM Bioinformatics ontology. Since Bio.Tools softwares are annotated with EDAZM in terms of topics and operations, the idea would be to navigate through training materials and their, associated softwares, to retrieve relevant EDAM annotations. \n",
    "\n",
    "**Results**: we build a knowledge graph (KG) leveraging i) the EDAM ontology, ii) BioSchemas tools and training material profiles, and iii) lifted content from Bio.Tools and TeSS. This KG can further be exploited with semantic queries (SPARQL), for instance we can search a training material based on bioinformatics definition or synonyms formalized in th EDAM ontology. \n",
    "\n",
    "Contact : alban.gaignard@univ-nantes.fr\n",
    "\n",
    "Work done during the [Elixir BioHackathon 2019](https://www.biohackathon-europe.org) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Transforming a TeSS entry to BioSchema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample JSON-LD data for the a TeSS training material\n",
      "\n",
      "{\n",
      "    \"@context\": {\n",
      "        \"edam\": \"http://edamontology.org/\",\n",
      "        \"hasPart\": \"schema:hasPart\",\n",
      "        \"name\": \"schema:name\",\n",
      "        \"schema\": \"http://schema.org/\"\n",
      "    },\n",
      "    \"@graph\": [\n",
      "        {\n",
      "            \"@id\": \"https://doi.org/10.5281/zenodo.3689221\",\n",
      "            \"@type\": \"schema:CreativeWork\",\n",
      "            \"hasPart\": {\n",
      "                \"@id\": \"https://bio.tools/tool/Data_Stewardship_Wizard\"\n",
      "            },\n",
      "            \"name\": \"Data Stewardship Wizard Workshop (Feb 2020)\"\n",
      "        }\n",
      "    ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "import requests\n",
    "import urllib3\n",
    "import pandas as pd\n",
    "from rdflib import ConjunctiveGraph\n",
    "urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)\n",
    "\n",
    "#url = 'https://tess.elixir-europe.org/materials.json_api?scientific_topics=RNA-Seq'\n",
    "url = 'https://tess.elixir-europe.org/materials.json_api'\n",
    "    \n",
    "r = requests.get(url)\n",
    "training = r.json()\n",
    "\n",
    "#print(json.dumps(training, indent=2))\n",
    "\n",
    "def rdfize_tess(item):\n",
    "    jsonld = {}\n",
    "    jsonld['@context'] = {\n",
    "        \"edam\": \"http://edamontology.org/\",\n",
    "        \"schema\": \"http://schema.org/\", \n",
    "        \"name\": \"schema:name\",\n",
    "        \"hasPart\": \"schema:hasPart\"\n",
    "    }\n",
    "    jsonld['@graph'] = []\n",
    "\n",
    "    #print(item['attributes']['title'])\n",
    "    #print()\n",
    "    #print(item['attributes'])\n",
    "    #print()\n",
    "    if item['attributes']['external-resources'] : \n",
    "        for extr in item['attributes']['external-resources']:\n",
    "            #print(extr)\n",
    "            if 'bio.tools' in extr['url']:\n",
    "                #print('\\t' + str(extr['url']))\n",
    "                tess_biotools_entry = {'@id':item['attributes']['url'], \n",
    "                 '@type': 'schema:CreativeWork', \n",
    "                 'name':item['attributes']['title'], \n",
    "                 'hasPart':{'@id':extr['url']}}\n",
    "                jsonld['@graph'].append(tess_biotools_entry)\n",
    "                \n",
    "    return json.dumps(jsonld, indent=4, sort_keys=True)\n",
    "\n",
    "print('Sample JSON-LD data for the a TeSS training material\\n')\n",
    "for item in training['data']:\n",
    "    print(rdfize_tess(item))\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Crawling the whole TeSS registry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15 % done\n",
      "29 % done\n",
      "44 % done\n",
      "59 % done\n",
      "74 % done\n",
      "88 % done\n",
      "666\n",
      "BioSchema output produced at tess.bioschema.ttl\n",
      "BioSchema output produced at tess.bioschema.jsonld\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "def crawl_tess(limit=-1):\n",
    "    graph = ConjunctiveGraph()\n",
    "    \n",
    "    \"\"\"\n",
    "    Go through all TeSS entries and print the tool home page if the keyword is found in the tool description.  \n",
    "    :param limit: an integer value specifying the max number of entries to be crawled, -1 by default, means no limit.\n",
    "    \"\"\"\n",
    "    \n",
    "    http = urllib3.PoolManager()\n",
    "    http.headers['Accept'] = 'application/json'\n",
    "    http.headers['Content-type'] = 'application/json'\n",
    "    \n",
    "    try:\n",
    "        req = http.request('GET', 'https://tess.elixir-europe.org/materials.json_api?page=1')\n",
    "        count_json = json.loads(req.data.decode('utf-8'))\n",
    "        #print(json.dumps(count_json, indent=2))\n",
    "        count = int(count_json['meta']['results-count'])\n",
    "        #print(str(count)+ \" available TeSS entries\")\n",
    "\n",
    "        i = 1\n",
    "        nb_tools = 1\n",
    "        has_next_page = True\n",
    "        while has_next_page :\n",
    "            #print('Crawling page '+str(i))\n",
    "            req = http.request('GET', 'https://tess.elixir-europe.org/materials.json_api?page=' + str(i))\n",
    "            try:\n",
    "                entry = json.loads(req.data.decode('utf-8'))\n",
    "            except JSONDecodeError as e:\n",
    "                print(\"Json decode error for \" + str(req.data.decode('utf-8')))\n",
    "                break\n",
    "            \n",
    "            #print(json.dumps(entry['links'], indent=2))\n",
    "            has_next_page = ('next' in entry['links'].keys())\n",
    "            \n",
    "\n",
    "            for tool in entry['data']:\n",
    "                jsonld = rdfize_tess(tool)\n",
    "                graph.parse(data=jsonld, format='json-ld')\n",
    "                #print(len(graph))\n",
    "                \n",
    "                #temp_graph = ConjunctiveGraph()\n",
    "                #temp_graph.parse(data=jsonld, format='json-ld')\n",
    "                #os.makedirs('./bio.tools.dataset/'+tool['biotoolsID'], exist_ok=True)\n",
    "                #temp_graph.serialize(format=\"json-ld\", \n",
    "                #                     auto_compact=True, \n",
    "                #                     destination=str('./bio.tools.dataset/'+tool['biotoolsID']+'/'+tool['biotoolsID']+'.jsonld'))\n",
    "                \n",
    "                nb_tools += 1\n",
    "                progress = nb_tools * 100 / count\n",
    "                if (nb_tools % 200 == 0) :\n",
    "                    print(str(round(progress))+\" % done\")\n",
    "                if ((limit != -1) and (nb_tools >= limit)):\n",
    "                    return graph\n",
    "            i += 1\n",
    "\n",
    "    except urllib3.exceptions.HTTPError as e:\n",
    "        print(e)\n",
    "    \n",
    "    return graph\n",
    "\n",
    "g = crawl_tess(limit=-1)\n",
    "print(len(g))\n",
    "g.serialize(format='turtle', destination='tess.bioschema.ttl')\n",
    "print('BioSchema output produced at tess.bioschema.ttl')\n",
    "g.serialize(format='json-ld', auto_compact=True, destination='tess.bioschema.jsonld')\n",
    "print('BioSchema output produced at tess.bioschema.jsonld')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Feeding the KG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "666\n",
      "196201\n",
      "229690\n",
      "CPU times: user 16.1 s, sys: 157 ms, total: 16.3 s\n",
      "Wall time: 16.4 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from rdflib import ConjunctiveGraph\n",
    "kg = ConjunctiveGraph()# 2. Crawling the whole TeSS registry\n",
    "kg.parse('tess.bioschema.ttl', format='turtle')\n",
    "print(len(kg))\n",
    "kg.parse('biotools.bioschema.ttl', format='turtle')\n",
    "print(len(kg))\n",
    "kg.parse('/Users/gaignard-a/Documents/Dev/fresh-toolbox/EDAM.owl')\n",
    "print(len(kg))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Querying the KG\n",
    "This query searches for training materials having as \"part-of\" relation some tools. The descriptions of the tools are then navigated to retreive their names, their associated topics, and the operations they perform.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Operation</th>\n",
       "      <th>Tool</th>\n",
       "      <th>Topic</th>\n",
       "      <th>Training material</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Computational biology</td>\n",
       "      <td>R / Bioconductor for everyone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Computational biology</td>\n",
       "      <td>R / Bioconductor for everyone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Computational biology</td>\n",
       "      <td>R / Bioconductor for everyone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>R / Bioconductor for everyone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>R / Bioconductor for everyone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>R / Bioconductor for everyone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Data management</td>\n",
       "      <td>R / Bioconductor for everyone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Data management</td>\n",
       "      <td>R / Bioconductor for everyone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Data management</td>\n",
       "      <td>R / Bioconductor for everyone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Statistics and probability</td>\n",
       "      <td>R / Bioconductor for everyone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Statistics and probability</td>\n",
       "      <td>R / Bioconductor for everyone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Statistics and probability</td>\n",
       "      <td>R / Bioconductor for everyone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Computational biology</td>\n",
       "      <td>Bioconductor for 'Omics Analysis (University o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Computational biology</td>\n",
       "      <td>Bioconductor for 'Omics Analysis (University o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Computational biology</td>\n",
       "      <td>Bioconductor for 'Omics Analysis (University o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>Bioconductor for 'Omics Analysis (University o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>Bioconductor for 'Omics Analysis (University o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>Bioconductor for 'Omics Analysis (University o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Data management</td>\n",
       "      <td>Bioconductor for 'Omics Analysis (University o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Data management</td>\n",
       "      <td>Bioconductor for 'Omics Analysis (University o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Data management</td>\n",
       "      <td>Bioconductor for 'Omics Analysis (University o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Statistics and probability</td>\n",
       "      <td>Bioconductor for 'Omics Analysis (University o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Statistics and probability</td>\n",
       "      <td>Bioconductor for 'Omics Analysis (University o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Statistics and probability</td>\n",
       "      <td>Bioconductor for 'Omics Analysis (University o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Computational biology</td>\n",
       "      <td>Working with Ranges infrastructure: annotating...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Computational biology</td>\n",
       "      <td>Working with Ranges infrastructure: annotating...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Computational biology</td>\n",
       "      <td>Working with Ranges infrastructure: annotating...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>Working with Ranges infrastructure: annotating...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>Working with Ranges infrastructure: annotating...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>Working with Ranges infrastructure: annotating...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2334</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Data management</td>\n",
       "      <td>R / Bioconductor for 'Omics Analysis (U. Idaho)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2335</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Data management</td>\n",
       "      <td>R / Bioconductor for 'Omics Analysis (U. Idaho)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2336</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Data management</td>\n",
       "      <td>R / Bioconductor for 'Omics Analysis (U. Idaho)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2337</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Statistics and probability</td>\n",
       "      <td>R / Bioconductor for 'Omics Analysis (U. Idaho)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2338</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Statistics and probability</td>\n",
       "      <td>R / Bioconductor for 'Omics Analysis (U. Idaho)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2339</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Statistics and probability</td>\n",
       "      <td>R / Bioconductor for 'Omics Analysis (U. Idaho)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2340</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Computational biology</td>\n",
       "      <td>Variant calling with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2341</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Computational biology</td>\n",
       "      <td>Variant calling with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2342</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Computational biology</td>\n",
       "      <td>Variant calling with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2343</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>Variant calling with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2344</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>Variant calling with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2345</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>Variant calling with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2346</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Data management</td>\n",
       "      <td>Variant calling with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2347</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Data management</td>\n",
       "      <td>Variant calling with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2348</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Data management</td>\n",
       "      <td>Variant calling with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2349</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Statistics and probability</td>\n",
       "      <td>Variant calling with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2350</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Statistics and probability</td>\n",
       "      <td>Variant calling with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2351</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Statistics and probability</td>\n",
       "      <td>Variant calling with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2352</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Computational biology</td>\n",
       "      <td>eQTL analysis -- an approach with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2353</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Computational biology</td>\n",
       "      <td>eQTL analysis -- an approach with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2354</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Computational biology</td>\n",
       "      <td>eQTL analysis -- an approach with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2355</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>eQTL analysis -- an approach with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2356</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>eQTL analysis -- an approach with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2357</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Bioinformatics</td>\n",
       "      <td>eQTL analysis -- an approach with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2358</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Data management</td>\n",
       "      <td>eQTL analysis -- an approach with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2359</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Data management</td>\n",
       "      <td>eQTL analysis -- an approach with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2360</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Data management</td>\n",
       "      <td>eQTL analysis -- an approach with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2361</th>\n",
       "      <td>Statistical calculation</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Statistics and probability</td>\n",
       "      <td>eQTL analysis -- an approach with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2362</th>\n",
       "      <td>Analysis</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Statistics and probability</td>\n",
       "      <td>eQTL analysis -- an approach with Bioconductor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2363</th>\n",
       "      <td>Data handling</td>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>Statistics and probability</td>\n",
       "      <td>eQTL analysis -- an approach with Bioconductor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2364 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Operation          Tool                       Topic  \\\n",
       "0     Statistical calculation  Bioconductor       Computational biology   \n",
       "1                    Analysis  Bioconductor       Computational biology   \n",
       "2               Data handling  Bioconductor       Computational biology   \n",
       "3     Statistical calculation  Bioconductor              Bioinformatics   \n",
       "4                    Analysis  Bioconductor              Bioinformatics   \n",
       "5               Data handling  Bioconductor              Bioinformatics   \n",
       "6     Statistical calculation  Bioconductor             Data management   \n",
       "7                    Analysis  Bioconductor             Data management   \n",
       "8               Data handling  Bioconductor             Data management   \n",
       "9     Statistical calculation  Bioconductor  Statistics and probability   \n",
       "10                   Analysis  Bioconductor  Statistics and probability   \n",
       "11              Data handling  Bioconductor  Statistics and probability   \n",
       "12    Statistical calculation  Bioconductor       Computational biology   \n",
       "13                   Analysis  Bioconductor       Computational biology   \n",
       "14              Data handling  Bioconductor       Computational biology   \n",
       "15    Statistical calculation  Bioconductor              Bioinformatics   \n",
       "16                   Analysis  Bioconductor              Bioinformatics   \n",
       "17              Data handling  Bioconductor              Bioinformatics   \n",
       "18    Statistical calculation  Bioconductor             Data management   \n",
       "19                   Analysis  Bioconductor             Data management   \n",
       "20              Data handling  Bioconductor             Data management   \n",
       "21    Statistical calculation  Bioconductor  Statistics and probability   \n",
       "22                   Analysis  Bioconductor  Statistics and probability   \n",
       "23              Data handling  Bioconductor  Statistics and probability   \n",
       "24    Statistical calculation  Bioconductor       Computational biology   \n",
       "25                   Analysis  Bioconductor       Computational biology   \n",
       "26              Data handling  Bioconductor       Computational biology   \n",
       "27    Statistical calculation  Bioconductor              Bioinformatics   \n",
       "28                   Analysis  Bioconductor              Bioinformatics   \n",
       "29              Data handling  Bioconductor              Bioinformatics   \n",
       "...                       ...           ...                         ...   \n",
       "2334  Statistical calculation  Bioconductor             Data management   \n",
       "2335                 Analysis  Bioconductor             Data management   \n",
       "2336            Data handling  Bioconductor             Data management   \n",
       "2337  Statistical calculation  Bioconductor  Statistics and probability   \n",
       "2338                 Analysis  Bioconductor  Statistics and probability   \n",
       "2339            Data handling  Bioconductor  Statistics and probability   \n",
       "2340  Statistical calculation  Bioconductor       Computational biology   \n",
       "2341                 Analysis  Bioconductor       Computational biology   \n",
       "2342            Data handling  Bioconductor       Computational biology   \n",
       "2343  Statistical calculation  Bioconductor              Bioinformatics   \n",
       "2344                 Analysis  Bioconductor              Bioinformatics   \n",
       "2345            Data handling  Bioconductor              Bioinformatics   \n",
       "2346  Statistical calculation  Bioconductor             Data management   \n",
       "2347                 Analysis  Bioconductor             Data management   \n",
       "2348            Data handling  Bioconductor             Data management   \n",
       "2349  Statistical calculation  Bioconductor  Statistics and probability   \n",
       "2350                 Analysis  Bioconductor  Statistics and probability   \n",
       "2351            Data handling  Bioconductor  Statistics and probability   \n",
       "2352  Statistical calculation  Bioconductor       Computational biology   \n",
       "2353                 Analysis  Bioconductor       Computational biology   \n",
       "2354            Data handling  Bioconductor       Computational biology   \n",
       "2355  Statistical calculation  Bioconductor              Bioinformatics   \n",
       "2356                 Analysis  Bioconductor              Bioinformatics   \n",
       "2357            Data handling  Bioconductor              Bioinformatics   \n",
       "2358  Statistical calculation  Bioconductor             Data management   \n",
       "2359                 Analysis  Bioconductor             Data management   \n",
       "2360            Data handling  Bioconductor             Data management   \n",
       "2361  Statistical calculation  Bioconductor  Statistics and probability   \n",
       "2362                 Analysis  Bioconductor  Statistics and probability   \n",
       "2363            Data handling  Bioconductor  Statistics and probability   \n",
       "\n",
       "                                      Training material  \n",
       "0                         R / Bioconductor for everyone  \n",
       "1                         R / Bioconductor for everyone  \n",
       "2                         R / Bioconductor for everyone  \n",
       "3                         R / Bioconductor for everyone  \n",
       "4                         R / Bioconductor for everyone  \n",
       "5                         R / Bioconductor for everyone  \n",
       "6                         R / Bioconductor for everyone  \n",
       "7                         R / Bioconductor for everyone  \n",
       "8                         R / Bioconductor for everyone  \n",
       "9                         R / Bioconductor for everyone  \n",
       "10                        R / Bioconductor for everyone  \n",
       "11                        R / Bioconductor for everyone  \n",
       "12    Bioconductor for 'Omics Analysis (University o...  \n",
       "13    Bioconductor for 'Omics Analysis (University o...  \n",
       "14    Bioconductor for 'Omics Analysis (University o...  \n",
       "15    Bioconductor for 'Omics Analysis (University o...  \n",
       "16    Bioconductor for 'Omics Analysis (University o...  \n",
       "17    Bioconductor for 'Omics Analysis (University o...  \n",
       "18    Bioconductor for 'Omics Analysis (University o...  \n",
       "19    Bioconductor for 'Omics Analysis (University o...  \n",
       "20    Bioconductor for 'Omics Analysis (University o...  \n",
       "21    Bioconductor for 'Omics Analysis (University o...  \n",
       "22    Bioconductor for 'Omics Analysis (University o...  \n",
       "23    Bioconductor for 'Omics Analysis (University o...  \n",
       "24    Working with Ranges infrastructure: annotating...  \n",
       "25    Working with Ranges infrastructure: annotating...  \n",
       "26    Working with Ranges infrastructure: annotating...  \n",
       "27    Working with Ranges infrastructure: annotating...  \n",
       "28    Working with Ranges infrastructure: annotating...  \n",
       "29    Working with Ranges infrastructure: annotating...  \n",
       "...                                                 ...  \n",
       "2334    R / Bioconductor for 'Omics Analysis (U. Idaho)  \n",
       "2335    R / Bioconductor for 'Omics Analysis (U. Idaho)  \n",
       "2336    R / Bioconductor for 'Omics Analysis (U. Idaho)  \n",
       "2337    R / Bioconductor for 'Omics Analysis (U. Idaho)  \n",
       "2338    R / Bioconductor for 'Omics Analysis (U. Idaho)  \n",
       "2339    R / Bioconductor for 'Omics Analysis (U. Idaho)  \n",
       "2340                  Variant calling with Bioconductor  \n",
       "2341                  Variant calling with Bioconductor  \n",
       "2342                  Variant calling with Bioconductor  \n",
       "2343                  Variant calling with Bioconductor  \n",
       "2344                  Variant calling with Bioconductor  \n",
       "2345                  Variant calling with Bioconductor  \n",
       "2346                  Variant calling with Bioconductor  \n",
       "2347                  Variant calling with Bioconductor  \n",
       "2348                  Variant calling with Bioconductor  \n",
       "2349                  Variant calling with Bioconductor  \n",
       "2350                  Variant calling with Bioconductor  \n",
       "2351                  Variant calling with Bioconductor  \n",
       "2352     eQTL analysis -- an approach with Bioconductor  \n",
       "2353     eQTL analysis -- an approach with Bioconductor  \n",
       "2354     eQTL analysis -- an approach with Bioconductor  \n",
       "2355     eQTL analysis -- an approach with Bioconductor  \n",
       "2356     eQTL analysis -- an approach with Bioconductor  \n",
       "2357     eQTL analysis -- an approach with Bioconductor  \n",
       "2358     eQTL analysis -- an approach with Bioconductor  \n",
       "2359     eQTL analysis -- an approach with Bioconductor  \n",
       "2360     eQTL analysis -- an approach with Bioconductor  \n",
       "2361     eQTL analysis -- an approach with Bioconductor  \n",
       "2362     eQTL analysis -- an approach with Bioconductor  \n",
       "2363     eQTL analysis -- an approach with Bioconductor  \n",
       "\n",
       "[2364 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "SELECT ?mat_label ?t_label ?op_label ?tool_name WHERE {\n",
    "    ?mat schema:hasPart ?tool ;\n",
    "        schema:name ?mat_label .\n",
    "    ?tool schema:applicationSubCategory ?topic ;\n",
    "       schema:featureList ?operation ;\n",
    "       schema:name ?tool_name .\n",
    "    ?topic rdfs:label ?t_label .\n",
    "    ?operation rdfs:label ?op_label .\n",
    "}\n",
    "\"\"\"\n",
    "res_data = []\n",
    "results = kg.query(query)\n",
    "for r in results:\n",
    "    #print(r)\n",
    "    res_data.append({'Training material':str(r['mat_label']), 'Tool':str(r['tool_name']), 'Topic':str(r['t_label']), 'Operation':str(r['op_label'])})   \n",
    "        \n",
    "df = pd.DataFrame(res_data)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
