{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# How to crawl bio.tools with its API ? "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook shows you how to automate the crawling of biotools to filter or transform its content. Please send any comment or feedback to alban.gaignard@univ-nantes.fr. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import urllib3\n",
    "import requests\n",
    "import json\n",
    "import argparse\n",
    "from argparse import RawTextHelpFormatter\n",
    "from json.decoder import JSONDecodeError\n",
    "import time\n",
    "import sys, os\n",
    "from rdflib import ConjunctiveGraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rdfize(json_entry):\n",
    "    \"\"\"\n",
    "    Transforms a biotools json entry into RDF, and returns a JSON-LD serialization. The following fields\n",
    "    are covered: contact, publication, EDAM topic, EDAM operation, EDAM inputs & outputs.\n",
    "    \"\"\"\n",
    "\n",
    "    entry = json_entry\n",
    "\n",
    "    try:\n",
    "\n",
    "        ctx = {\n",
    "            \"@context\": {\n",
    "                \"@base\": \"https://bio.tools/\",\n",
    "                \"biotools\": \"https://bio.tools/ontology/\",\n",
    "                \"edam\": \"http://edamontology.org/\",\n",
    "                \"pubmed\": \"https://www.ncbi.nlm.nih.gov/pubmed/\",\n",
    "                \"pmc\": \"https://www.ncbi.nlm.nih.gov/pmc/\",\n",
    "                \"doi\": \"https://doi.org/\",\n",
    "                \"dc\": \"http://purl.org/dc/terms/\",\n",
    "                \"rdfs\": \"http://www.w3.org/2000/01/rdf-schema#\",\n",
    "                \n",
    "                \"sc\": \"http://schema.org/\", \n",
    "                \"bsc\": \"http://bioschemas.org/\",\n",
    "                \n",
    "                #\"description\": {'@id' : 'sc:description','@type' : 'xsd:string'},\n",
    "                #\"description\": {'@id' : 'sc:description','@type' : 'sc:Text'},\n",
    "                \"description\": 'sc:description',\n",
    "                \"name\": \"sc:name\",\n",
    "                \"homepage\": \"sc:url\",\n",
    "                \"toolType\": 'sc:additionalType',\n",
    "                \"hasTopic\": 'sc:applicationSubCategory',\n",
    "                \"author\": 'sc:author',\n",
    "                \"hasPublication\": \"sc:citation\",\n",
    "                \"hasOperation\": \"sc:featureList\",\n",
    "                \"license\": \"sc:license\",\n",
    "                \"version\": \"sc:version\",\n",
    "\n",
    "                # \"hasContact\": \"dc:publisher\",\n",
    "                \n",
    "\n",
    "                # \"id\": \"datacite:identifier\",\n",
    "                #\"id\": \"dc:identifier\",\n",
    "                # \"name\": \"datacite:title\",\n",
    "                \n",
    "                # \"description\": \"datacite:description\",\n",
    "                \n",
    "                \n",
    "                #\"license\": \"dc:license\",\n",
    "                #\"hasContact\": \"datacite:contributor\",\n",
    "                #\"toolType\": \"datacite:resourceType\",\n",
    "                #\"additionDate\": \"datacite:date\",\n",
    "                #\"language\": \"datacite:format\",\n",
    "                \n",
    "                #\"download\": \"datacite:alternateIdentifier\",\n",
    "\n",
    "                \"hasInputData\": \"edam:has_input\",\n",
    "                \"hasOutputData\": \"edam:has_output\",\n",
    "            }\n",
    "        }\n",
    "        entry.update(ctx)\n",
    "        \n",
    "        entry['@id'] = str(entry['biotoolsID'])\n",
    "        #entry['@type'] = ['bsc:Tool','sc:SoftwareApplication']\n",
    "        entry['@type'] = ['sc:SoftwareApplication']\n",
    "        entry['applicationCategory'] = 'Computational science tool'\n",
    "\n",
    "        # for contact in entry['contact']:\n",
    "        #     if not \"hasContact\" in entry.keys():\n",
    "        #         entry['hasContact'] = [contact['name']]\n",
    "        #     else :\n",
    "        #         entry['hasContact'].append(contact['name'])\n",
    "\n",
    "        # for download in entry['download']:\n",
    "        #     if download['url']:\n",
    "        #         if not \"download\" in entry.keys():\n",
    "        #             entry['download'] = [download['url']]\n",
    "        #         else :\n",
    "        #             entry['download'].append(download['url'])\n",
    "        \n",
    "        #for toolType in entry['toolType']:\n",
    "        \n",
    "        for credit in entry['credit']:\n",
    "            if credit['name']:\n",
    "                if not 'author' in entry.keys():\n",
    "                    entry['author'] = [credit['name']]\n",
    "                else:\n",
    "                    entry['author'].append(credit['name'])\n",
    "            \n",
    "        for publication in entry['publication']:\n",
    "            if publication['pmid']:\n",
    "                if not \"hasPublication\" in entry.keys():\n",
    "                    #entry['hasPublication'] = [{\"@id\": 'pubmed:' + publication['pmid']}]\n",
    "                    entry['hasPublication'] = ['pubmed:' + publication['pmid']]\n",
    "                else:\n",
    "                    #entry['hasPublication'].append({\"@id\": 'pubmed:' + publication['pmid']})\n",
    "                    entry['hasPublication'].append('pubmed:' + publication['pmid'])\n",
    "            if publication['pmcid']:\n",
    "                if not \"hasPublication\" in entry.keys():\n",
    "                    entry['hasPublication'] = ['pmcid:' + publication['pmcid']]\n",
    "                else:\n",
    "                    entry['hasPublication'].append('pmcid:' + publication['pmcid'])\n",
    "            if publication['doi']:\n",
    "                if not (\"<\" in publication['doi'] or \">\" in publication['doi']):\n",
    "                    if not \"hasPublication\" in entry.keys():\n",
    "                        entry['hasPublication'] = [{\"@id\": \"https://doi.org/\" + publication['doi'], \"@type\":\"sc:CreativeWork\"}]\n",
    "                    else:\n",
    "                        entry['hasPublication'].append({\"@id\": \"https://doi.org/\" + publication['doi'], \"@type\":\"sc:CreativeWork\"})\n",
    "\n",
    "        for item in entry['function']:\n",
    "            for op in item['operation']:\n",
    "                if not \"hasOperation\" in entry.keys():\n",
    "                    entry['hasOperation'] = [{\"@id\": op['uri']}]\n",
    "                else:\n",
    "                    entry['hasOperation'].append({\"@id\": op['uri']})\n",
    "\n",
    "            for input in item['input']:\n",
    "                if not \"hasInputData\" in entry.keys():\n",
    "                    entry['hasInputData'] = [{\"@id\": input['data']['uri']}]\n",
    "                else:\n",
    "                    entry['hasInputData'].append({\"@id\": input['data']['uri']})\n",
    "\n",
    "            for output in item['output']:\n",
    "                if not \"hasOutputData\" in entry.keys():\n",
    "                    entry['hasOutputData'] = [{\"@id\": output['data']['uri']}]\n",
    "                else:\n",
    "                    entry['hasOutputData'].append({\"@id\": output['data']['uri']})\n",
    "\n",
    "        for item in entry['topic']:\n",
    "            if not \"hasTopic\" in entry.keys():\n",
    "                entry['hasTopic'] = [{\"@id\": item['uri']}]\n",
    "            else:\n",
    "                entry['hasTopic'].append({\"@id\": item['uri']})\n",
    "\n",
    "    except KeyError as error:\n",
    "        print(json.dumps(entry, indent=4, sort_keys=True))\n",
    "        print()\n",
    "        \n",
    "    #print(json.dumps(entry, indent=4, sort_keys=True))\n",
    "\n",
    "    raw_jld = json.dumps(entry, indent=4, sort_keys=True)\n",
    "    return raw_jld"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)\n",
    "\n",
    "\n",
    "\n",
    "def crawl_biotools(keyword, limit=-1):\n",
    "    graph = ConjunctiveGraph()\n",
    "    \n",
    "    \"\"\"\n",
    "    Go through all bio.tools entries and print the tool home page if the keyword is found in the tool description.  \n",
    "    :param limit: an integer value specifying the max number of entries to be crawled, -1 by default, means no limit.\n",
    "    \"\"\"\n",
    "    \n",
    "    http = urllib3.PoolManager()\n",
    "    http.headers['Accept'] = 'application/json'\n",
    "    http.headers['Content-type'] = 'application/json'\n",
    "    \n",
    "    try:\n",
    "        req = http.request('GET', 'https://bio.tools/api/tool/?page=1&?format=json')\n",
    "        count_json = json.loads(req.data.decode('utf-8'))\n",
    "        count = int(count_json['count'])\n",
    "        print(str(count)+ \" available BioTools entries\")\n",
    "\n",
    "        i = 1\n",
    "        nb_tools = 1\n",
    "        has_next_page = True\n",
    "        while has_next_page :\n",
    "            req = http.request('GET', 'https://bio.tools/api/tool/?page=' + str(i) + '&?format=json')\n",
    "            try:\n",
    "                entry = json.loads(req.data.decode('utf-8'))\n",
    "            except JSONDecodeError as e:\n",
    "                print(\"Json decode error for \" + str(req.data.decode('utf-8')))\n",
    "                break\n",
    "            has_next_page = (entry['next'] != None)\n",
    "\n",
    "            for tool in entry['list']:\n",
    "                jsonld = rdfize(tool)\n",
    "                graph.parse(data=jsonld, format='json-ld')\n",
    "                \n",
    "                nb_tools += 1\n",
    "                progress = nb_tools * 100 / count\n",
    "                if (nb_tools % 500 == 0) :\n",
    "                    print(str(round(progress))+\" % done\")\n",
    "                if ((limit != -1) and (nb_tools >= limit)):\n",
    "                    return graph\n",
    "            i += 1\n",
    "\n",
    "    except urllib3.exceptions.HTTPError as e:\n",
    "        print(e)\n",
    "    \n",
    "    return graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we just crawl the first `1000` entries and search for tools with the `rare disease` keyword in their description field. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#crawl_biotools(\"snp\", limit=1)\n",
    "#len(graph)\n",
    "#print(graph.serialize(format='turtle').decode())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"@context\": {\n",
      "        \"@base\": \"https://bio.tools/\",\n",
      "        \"author\": \"sc:author\",\n",
      "        \"biotools\": \"https://bio.tools/ontology/\",\n",
      "        \"bsc\": \"http://bioschemas.org/\",\n",
      "        \"dc\": \"http://purl.org/dc/terms/\",\n",
      "        \"description\": \"sc:description\",\n",
      "        \"doi\": \"https://doi.org/\",\n",
      "        \"edam\": \"http://edamontology.org/\",\n",
      "        \"hasInputData\": \"edam:has_input\",\n",
      "        \"hasOperation\": \"sc:featureList\",\n",
      "        \"hasOutputData\": \"edam:has_output\",\n",
      "        \"hasPublication\": \"sc:citation\",\n",
      "        \"hasTopic\": \"sc:applicationSubCategory\",\n",
      "        \"homepage\": \"sc:url\",\n",
      "        \"license\": \"sc:license\",\n",
      "        \"name\": \"sc:name\",\n",
      "        \"pmc\": \"https://www.ncbi.nlm.nih.gov/pmc/\",\n",
      "        \"pubmed\": \"https://www.ncbi.nlm.nih.gov/pubmed/\",\n",
      "        \"rdfs\": \"http://www.w3.org/2000/01/rdf-schema#\",\n",
      "        \"sc\": \"http://schema.org/\",\n",
      "        \"toolType\": \"sc:additionalType\",\n",
      "        \"version\": \"sc:version\"\n",
      "    },\n",
      "    \"@id\": \"rsat\",\n",
      "    \"@type\": [\n",
      "        \"sc:SoftwareApplication\"\n",
      "    ],\n",
      "    \"accessibility\": [],\n",
      "    \"additionDate\": \"2017-02-10T11:35:21Z\",\n",
      "    \"applicationCategory\": \"Computational science tool\",\n",
      "    \"author\": [\n",
      "        \"Jacques van Helden\"\n",
      "    ],\n",
      "    \"biotoolsCURIE\": \"biotools:rsat\",\n",
      "    \"biotoolsID\": \"rsat\",\n",
      "    \"collectionID\": [\n",
      "        \"Animal and Crop Genomics\"\n",
      "    ],\n",
      "    \"confidence_flag\": null,\n",
      "    \"cost\": null,\n",
      "    \"credit\": [\n",
      "        {\n",
      "            \"email\": \"Jacques.van-Helden@univ-amu.fr\",\n",
      "            \"name\": \"Jacques van Helden\",\n",
      "            \"note\": null,\n",
      "            \"orcidid\": null,\n",
      "            \"typeEntity\": \"Person\",\n",
      "            \"typeRole\": [\n",
      "                \"Primary contact\"\n",
      "            ],\n",
      "            \"url\": null\n",
      "        }\n",
      "    ],\n",
      "    \"description\": \"Regulatory Sequence Analysis Tools detects regulatory signals in non-coding sequences, tools include sequence retrieval, pattern discovery, pattern matching, genome-scale pattern matching, feature-map drawing, random sequence generation and other utilities. Tools may be accessed separately or connected to other tools. As well, the tools are available as web services for integration into programmatic work flows.\",\n",
      "    \"documentation\": [\n",
      "        {\n",
      "            \"note\": null,\n",
      "            \"type\": \"General\",\n",
      "            \"url\": \"http://rsat.eu\"\n",
      "        }\n",
      "    ],\n",
      "    \"download\": [],\n",
      "    \"editPermission\": {\n",
      "        \"authors\": [\n",
      "            \"\"\n",
      "        ],\n",
      "        \"type\": \"group\"\n",
      "    },\n",
      "    \"elixirNode\": [],\n",
      "    \"elixirPlatform\": [],\n",
      "    \"elixir_badge\": 0,\n",
      "    \"function\": [\n",
      "        {\n",
      "            \"cmd\": null,\n",
      "            \"input\": [],\n",
      "            \"note\": null,\n",
      "            \"operation\": [\n",
      "                {\n",
      "                    \"term\": \"Sequence motif discovery\",\n",
      "                    \"uri\": \"http://edamontology.org/operation_0238\"\n",
      "                },\n",
      "                {\n",
      "                    \"term\": \"Data retrieval\",\n",
      "                    \"uri\": \"http://edamontology.org/operation_2422\"\n",
      "                },\n",
      "                {\n",
      "                    \"term\": \"Transcriptional regulatory element prediction\",\n",
      "                    \"uri\": \"http://edamontology.org/operation_0438\"\n",
      "                },\n",
      "                {\n",
      "                    \"term\": \"cis-regulatory element prediction\",\n",
      "                    \"uri\": \"http://edamontology.org/operation_0441\"\n",
      "                }\n",
      "            ],\n",
      "            \"output\": []\n",
      "        }\n",
      "    ],\n",
      "    \"hasOperation\": [\n",
      "        {\n",
      "            \"@id\": \"http://edamontology.org/operation_0238\"\n",
      "        },\n",
      "        {\n",
      "            \"@id\": \"http://edamontology.org/operation_2422\"\n",
      "        },\n",
      "        {\n",
      "            \"@id\": \"http://edamontology.org/operation_0438\"\n",
      "        },\n",
      "        {\n",
      "            \"@id\": \"http://edamontology.org/operation_0441\"\n",
      "        }\n",
      "    ],\n",
      "    \"hasPublication\": [\n",
      "        \"pubmed:10641039\",\n",
      "        \"pubmed:12824373\",\n",
      "        \"pubmed:9719638\",\n",
      "        \"pubmed:18495751\"\n",
      "    ],\n",
      "    \"hasTopic\": [\n",
      "        {\n",
      "            \"@id\": \"http://edamontology.org/topic_0621\"\n",
      "        },\n",
      "        {\n",
      "            \"@id\": \"http://edamontology.org/topic_0749\"\n",
      "        },\n",
      "        {\n",
      "            \"@id\": \"http://edamontology.org/topic_0203\"\n",
      "        },\n",
      "        {\n",
      "            \"@id\": \"http://edamontology.org/topic_3512\"\n",
      "        },\n",
      "        {\n",
      "            \"@id\": \"http://edamontology.org/topic_3125\"\n",
      "        }\n",
      "    ],\n",
      "    \"homepage\": \"http://rsat.eu\",\n",
      "    \"homepage_status\": 0,\n",
      "    \"language\": [\n",
      "        \"Java\",\n",
      "        \"Perl\",\n",
      "        \"Python\"\n",
      "    ],\n",
      "    \"lastUpdate\": \"2019-06-19T13:43:40Z\",\n",
      "    \"license\": \"Other\",\n",
      "    \"link\": [],\n",
      "    \"maturity\": null,\n",
      "    \"name\": \"RSAT\",\n",
      "    \"operatingSystem\": [\n",
      "        \"Linux\",\n",
      "        \"Windows\",\n",
      "        \"Mac\"\n",
      "    ],\n",
      "    \"otherID\": [],\n",
      "    \"owner\": \"jvanheld\",\n",
      "    \"publication\": [\n",
      "        {\n",
      "            \"doi\": null,\n",
      "            \"metadata\": {\n",
      "                \"abstract\": \"A series of computer programs were developed for the analysis of regulatory sequences, with a special focus on yeast. These tools are publicly available on the web (http://www.ucmb.ulb.ac.be/bioinformatics/rsa-tools/). Basically, three classical problems can be addressed: (a) search for known regulatory patterns in the upstream regions of known genes; (b) discovery of unknown regulatory patterns within a set of upstream regions known to be co-regulated; (c) search for unknown genes potentially regulated by a known transcription factor. Each of these tasks can be performed on basis of a simple (string) or more refined (matrix) description of the regulatory patterns. A feature-map progam automatically generates visual representations of the positions at which patterns were found. The site also provides a series of general utilities, such as generation of random sequence, automatic drawing of XY graphs, interconversions between sequence formats, etc. Several tools are linked together to allow their sequential utilization (piping), but each one can also be used independently by filling the web form with external data. This widens the scope of the site to the analysis of nonregulatory and/or non-yeast sequences. Copyright (C) 2000 John Wiley and Sons, Ltd.\",\n",
      "                \"authors\": [\n",
      "                    {\n",
      "                        \"name\": \"Van Helden J.\"\n",
      "                    },\n",
      "                    {\n",
      "                        \"name\": \"Andre B.\"\n",
      "                    },\n",
      "                    {\n",
      "                        \"name\": \"Collado-Vides J.\"\n",
      "                    }\n",
      "                ],\n",
      "                \"citationCount\": 145,\n",
      "                \"date\": \"2000-01-30T00:00:00Z\",\n",
      "                \"journal\": \"Yeast\",\n",
      "                \"title\": \"A web site for the computational analysis of yeast regulatory sequences\"\n",
      "            },\n",
      "            \"pmcid\": null,\n",
      "            \"pmid\": \"10641039\",\n",
      "            \"type\": null,\n",
      "            \"version\": null\n",
      "        },\n",
      "        {\n",
      "            \"doi\": null,\n",
      "            \"metadata\": {\n",
      "                \"abstract\": \"The web resource Regulatory Sequence Analysis Tools (RSAT) (http://rsat.ulb.ac.be/rsat) offers a collection of software tools dedicated to the prediction of regulatory sites in non-coding DNA sequences. These tools include sequence retrieval, pattern discovery, pattern matching, genome-scale pattern matching, feature-map drawing, random sequence generation and other utilities. Alternative formats are supported for the representation of regulatory motifs (strings or position-specific scoring matrices) and several algorithms are proposed for pattern discovery. RSAT currently holds >100 fully sequenced genomes and these data are regularly updated from GenBank.\",\n",
      "                \"authors\": [\n",
      "                    {\n",
      "                        \"name\": \"van Helden J.\"\n",
      "                    }\n",
      "                ],\n",
      "                \"citationCount\": 321,\n",
      "                \"date\": \"2003-07-01T00:00:00Z\",\n",
      "                \"journal\": \"Nucleic Acids Research\",\n",
      "                \"title\": \"Regulatory Sequence Analysis Tools\"\n",
      "            },\n",
      "            \"pmcid\": null,\n",
      "            \"pmid\": \"12824373\",\n",
      "            \"type\": null,\n",
      "            \"version\": null\n",
      "        },\n",
      "        {\n",
      "            \"doi\": null,\n",
      "            \"metadata\": {\n",
      "                \"abstract\": \"We present here a simple and fast method allowing the isolation of DNA binding sites for transcription factors from families of coregulated genes, with results illustrated in Saccharomyces cerevisiae. Although conceptually simple, the algorithm proved efficient for extracting, from most of the yeast regulatory families analyzed, the upstream regulatory sequences which had been previously found by experimental analysis. Furthermore, putative new regulatory sites are predicted within upstream regions of several regulons. The method is based on the detection of over-represented oligonucleotides. A specificity of this approach is to define the statistical significance of a site based on tables of oligonucleotide frequencies observed in all non-coding sequences from the yeast genome. In contrast with heuristic methods, this oligonucleotide analysis is rigorous and exhaustive. Its range of detection is however limited to relatively simple patterns: short motifs with a highly conserved core. These features seem to be shared by a good number of regulatory sites in yeast. This, and similar methods, should be increasingly required to identify unknown regulatory elements within the numerous new coregulated families resulting from measurements of gene expression levels at the genomic scale.\",\n",
      "                \"authors\": [\n",
      "                    {\n",
      "                        \"name\": \"Van Helden J.\"\n",
      "                    },\n",
      "                    {\n",
      "                        \"name\": \"Andre B.\"\n",
      "                    },\n",
      "                    {\n",
      "                        \"name\": \"Collado-Vides J.\"\n",
      "                    }\n",
      "                ],\n",
      "                \"citationCount\": 563,\n",
      "                \"date\": \"1998-09-04T00:00:00Z\",\n",
      "                \"journal\": \"Journal of Molecular Biology\",\n",
      "                \"title\": \"Extracting regulatory sites from the upstream region of yeast genes by computational analysis of oligonucleotide frequencies\"\n",
      "            },\n",
      "            \"pmcid\": null,\n",
      "            \"pmid\": \"9719638\",\n",
      "            \"type\": null,\n",
      "            \"version\": null\n",
      "        },\n",
      "        {\n",
      "            \"doi\": null,\n",
      "            \"metadata\": {\n",
      "                \"abstract\": \"The regulatory sequence analysis tools (RSAT, http://rsat.ulb.ac.be/rsat/) is a software suite that integrates a wide collection of modular tools for the detection of cis-regulatory elements in genome sequences. The suite includes programs for sequence retrieval, pattern discovery, phylogenetic footprint detection, pattern matching, genome scanning and feature map drawing. Random controls can be performed with random gene selections or by generating random sequences according to a variety of background models (Bernoulli, Markov). Beyond the original word-based pattern-discovery tools (oligo-analysis and dyad-analysis), we recently added a battery of tools for matrix-based detection of cis-acting elements, with some original features (adaptive background models, Markov-chain estimation of P-values) that do not exist in other matrix-based scanning tools. The web server offers an intuitive interface, where each program can be accessed either separately or connected to the other tools. In addition, the tools are now available as web services, enabling their integration in programmatic workflows. Genomes are regularly updated from various genome repositories (NCBI and EnsEMBL) and 682 organisms are currently supported. Since 1998, the tools have been used by several hundreds of researchers from all over the world. Several predictions made with RSAT were validated experimentally and published.\",\n",
      "                \"authors\": [\n",
      "                    {\n",
      "                        \"name\": \"Thomas-Chollier M.\"\n",
      "                    },\n",
      "                    {\n",
      "                        \"name\": \"Sand O.\"\n",
      "                    },\n",
      "                    {\n",
      "                        \"name\": \"Turatsinze J.V.\"\n",
      "                    },\n",
      "                    {\n",
      "                        \"name\": \"Janky R.\"\n",
      "                    },\n",
      "                    {\n",
      "                        \"name\": \"Defrance M.\"\n",
      "                    },\n",
      "                    {\n",
      "                        \"name\": \"Vervisch E.\"\n",
      "                    },\n",
      "                    {\n",
      "                        \"name\": \"Brohee S.\"\n",
      "                    },\n",
      "                    {\n",
      "                        \"name\": \"van Helden J.\"\n",
      "                    }\n",
      "                ],\n",
      "                \"citationCount\": 195,\n",
      "                \"date\": \"2008-01-01T00:00:00Z\",\n",
      "                \"journal\": \"Nucleic acids research\",\n",
      "                \"title\": \"RSAT: regulatory sequence analysis tools.\"\n",
      "            },\n",
      "            \"pmcid\": null,\n",
      "            \"pmid\": \"18495751\",\n",
      "            \"type\": null,\n",
      "            \"version\": null\n",
      "        }\n",
      "    ],\n",
      "    \"relation\": [],\n",
      "    \"toolType\": [\n",
      "        \"Web application\",\n",
      "        \"Suite\"\n",
      "    ],\n",
      "    \"topic\": [\n",
      "        {\n",
      "            \"term\": \"Model organisms\",\n",
      "            \"uri\": \"http://edamontology.org/topic_0621\"\n",
      "        },\n",
      "        {\n",
      "            \"term\": \"Transcription factors and regulatory sites\",\n",
      "            \"uri\": \"http://edamontology.org/topic_0749\"\n",
      "        },\n",
      "        {\n",
      "            \"term\": \"Gene expression\",\n",
      "            \"uri\": \"http://edamontology.org/topic_0203\"\n",
      "        },\n",
      "        {\n",
      "            \"term\": \"Gene transcripts\",\n",
      "            \"uri\": \"http://edamontology.org/topic_3512\"\n",
      "        },\n",
      "        {\n",
      "            \"term\": \"DNA binding sites\",\n",
      "            \"uri\": \"http://edamontology.org/topic_3125\"\n",
      "        }\n",
      "    ],\n",
      "    \"validated\": 0,\n",
      "    \"version\": []\n",
      "}\n",
      "\n",
      "@prefix biotools: <https://bio.tools/ontology/> .\n",
      "@prefix bsc: <http://bioschemas.org/> .\n",
      "@prefix dc: <http://purl.org/dc/terms/> .\n",
      "@prefix doi: <https://doi.org/> .\n",
      "@prefix edam: <http://edamontology.org/> .\n",
      "@prefix pmc: <https://www.ncbi.nlm.nih.gov/pmc/> .\n",
      "@prefix pubmed: <https://www.ncbi.nlm.nih.gov/pubmed/> .\n",
      "@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .\n",
      "@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .\n",
      "@prefix sc: <http://schema.org/> .\n",
      "@prefix xml: <http://www.w3.org/XML/1998/namespace> .\n",
      "@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .\n",
      "\n",
      "<https://bio.tools/rsat> a sc:SoftwareApplication ;\n",
      "    sc:additionalType \"Suite\",\n",
      "        \"Web application\" ;\n",
      "    sc:applicationSubCategory edam:topic_0203,\n",
      "        edam:topic_0621,\n",
      "        edam:topic_0749,\n",
      "        edam:topic_3125,\n",
      "        edam:topic_3512 ;\n",
      "    sc:author \"Jacques van Helden\" ;\n",
      "    sc:citation \"pubmed:10641039\",\n",
      "        \"pubmed:12824373\",\n",
      "        \"pubmed:18495751\",\n",
      "        \"pubmed:9719638\" ;\n",
      "    sc:description \"Regulatory Sequence Analysis Tools detects regulatory signals in non-coding sequences, tools include sequence retrieval, pattern discovery, pattern matching, genome-scale pattern matching, feature-map drawing, random sequence generation and other utilities. Tools may be accessed separately or connected to other tools. As well, the tools are available as web services for integration into programmatic work flows.\" ;\n",
      "    sc:featureList edam:operation_0238,\n",
      "        edam:operation_0438,\n",
      "        edam:operation_0441,\n",
      "        edam:operation_2422 ;\n",
      "    sc:license \"Other\" ;\n",
      "    sc:name \"RSAT\" ;\n",
      "    sc:url \"http://rsat.eu\" .\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "url = 'https://bio.tools/api/tool/rsat?format=json'\n",
    "#url = 'https://bio.tools/api/tool/3SRP?format=json'\n",
    "\n",
    "r = requests.get(url)\n",
    "tool = r.json()\n",
    "print(rdfize(tool))\n",
    "\n",
    "g = ConjunctiveGraph()\n",
    "g.parse(data = rdfize(tool), format=\"json-ld\")\n",
    "print()\n",
    "print(g.serialize(format=\"turtle\").decode())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14767 available BioTools entries\n",
      "3 % done\n",
      "7 % done\n",
      "12537\n",
      "BioSchema output produced at biotools.bioschema.ttl\n",
      "BioSchema output produced at biotools.bioschema.json-ld\n",
      "CPU times: user 4.03 s, sys: 114 ms, total: 4.14 s\n",
      "Wall time: 15.4 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "g = crawl_biotools(\"\", limit=10)\n",
    "print(len(g))\n",
    "g.serialize(format='turtle', destination='biotools.bioschema.ttl')\n",
    "print('BioSchema output produced at biotools.bioschema.ttl')\n",
    "g.serialize(format='json-ld', destination='biotools.bioschema.json-ld')\n",
    "print('BioSchema output produced at biotools.bioschema.json-ld')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validation of data shapes (SHACL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
